HS-1302: Move detectors to be part of NodeScan

cgorantla opened the pull request:
## Description
Currently Inventory service sends detectors request after NodeScan. We can combine both steps.

## Jira link(s)
- https://issues.opennms.org/browse/HS-1302

## Flagged for review
<!-- Flag things as "needs a close look" for reviewers, if necessary. Include as much detail as possible (line numbers, concerns, and so on). -->

## Checklist
* [ ] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts

cgorantla commented on shared-lib/protobuf/src/main/proto/types.proto:
```diff
@@ -0,0 +1,40 @@
+option java_multiple_files = true;
+option java_package = "org.opennms.inventory.types";
+
+enum ServiceType {
```
Plan is to replace MonitorType with ServiceType, will make subsequent changes to replace MonitorType in another PR

yangli2014 reviewed (commented):
I got the following error when I added a new node
```
14:32:04.502 ERROR [service-#86] Error while node scan
java.lang.IllegalStateException: stream has already been operated upon or closed
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:246)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
	at org.opennms.horizon.minion.nodescan.NodeScanner.lambda$scan$1(NodeScanner.java:108)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	
```

cgorantla commented:
> I got the following error when I added a new node
> 
> ```
> 14:32:04.502 ERROR [service-#86] Error while node scan
> java.lang.IllegalStateException: stream has already been operated upon or closed
> 	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:246)
> 	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
> 	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
> 	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
> 	at org.opennms.horizon.minion.nodescan.NodeScanner.lambda$scan$1(NodeScanner.java:108)
> 	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
> 	
> ```

Can you please fetch the latest updates in the branch and try again ?

cgorantla commented:
> I have three SNMP devices: 10.244.0.10, 10.244.0.11, and 10.244.0.17. I added a discovery with IP rang 10.244.0.10-17 and all of these three devices were detected with extra devices. Hower only one device get SNMP scan results and changed the node label. I found the following error in the inventory log.
> 
> ```
> 2023-03-31T18:54:58.221Z  INFO 1 --- [ntainer#2-0-C-1] o.o.h.i.s.t.r.DetectorResponseService    : ICMP not detected on ip address = [-2, -128, 0, 0, 0, 0, 0, 0, 40, 124, 121, -1, -2, -9, -92, -122]
> 2023-03-31T18:54:58.221Z  INFO 1 --- [ntainer#2-0-C-1] o.o.h.i.s.t.r.DetectorResponseService    : Received Detector Response = ip_address: "0:0:0:0:0:0:0:1"
> service: SNMP
>  for tenant = opennms-prime and location = Default
> 2023-03-31T18:54:58.257Z ERROR 1 --- [ntainer#2-0-C-1] o.o.h.i.c.TaskSetResultsConsumer         : Error while processing kafka message for TaskResults: 
> 
> org.springframework.dao.IncorrectResultSizeDataAccessException: query did not return a unique result: 2
> 	at org.springframework.orm.jpa.EntityManagerFactoryUtils.convertJpaAccessExceptionIfPossible(EntityManagerFactoryUtils.java:382) ~[spring-orm-6.0.3.jar:6.0.3]
> 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:235) ~[spring-orm-6.0.3.jar:6.0.3]
> 	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:550) ~[spring-orm-6.0.3.jar:6.0.3]
> 	at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.0.3.jar:6.0.3]
> 	at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:242) ~[spring-tx-6.0.3.jar:6.0.3]
> 	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:152) ~[spring-tx-6.0.3.jar:6.0.3]
> 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.0.3.jar:6.0.3]
> 	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:134) ~[spring-data-jpa-3.0.0.jar:3.0.0]
> 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.0.3.jar:6.0.3]
> 	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97) ~[spring-aop-6.0.3.jar:6.0.3]
> 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.0.3.jar:6.0.3]
> 	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:218) ~[spring-aop-6.0.3.jar:6.0.3]
> 	at jdk.proxy4/jdk.proxy4.$Proxy156.findByIpAddressAndLocationAndTenantId(Unknown Source) ~[na:na]
> 	at org.opennms.horizon.inventory.service.taskset.response.DetectorResponseService.processDetectorResults(DetectorResponseService.java:108) ~[classes/:na]
> 	at org.opennms.horizon.inventory.service.taskset.response.ScannerResponseService.lambda$processNodeScanResponse$2(ScannerResponseService.java:215) ~[classes/:an]
> ```

Thanks for validating this.  I think 06caa1343cf5b934c65f45afbc19d7907f6b3b10 should fix this. When adding nodes, this checks for all existing ip interfaces instead of just the primary interface.

cgorantla commented:
currently ICMP discovery parsing doesn't support parsing IP ranges other than  standard IP Address ex: 192.168.1.1-192.168.1255, Created story to support other format such as cidr https://opennms.atlassian.net/browse/HS-1373.
